precompiles/dapps-staking/src/mock.rs

```diff
@@ -0,0 +1,418 @@
+    type Header = Header;
+    type Event = Event;
+    type BlockHashCount = BlockHashCount;
+    type DbWeight = ();
+    type Version = ();
+    type PalletInfo = PalletInfo;
+    type AccountData = pallet_balances::AccountData<Balance>;
+    type OnNewAccount = ();
+    type OnKilledAccount = ();
+    type SystemWeightInfo = ();
+    type SS58Prefix = ();
+    type OnSetCode = ();
+}
+
+parameter_types! {
+    pub const ExistentialDeposit: u128 = 1;
+}
+impl pallet_balances::Config for TestRuntime {
+    type MaxReserves = ();
+    type ReserveIdentifier = [u8; 4];
+    type MaxLocks = ();
+    type Balance = Balance;
+    type Event = Event;
+    type DustRemoval = ();
+    type ExistentialDeposit = ExistentialDeposit;
+    type AccountStore = System;
+    type WeightInfo = ();
+}
+
+pub fn precompile_address() -> H160 {
+    H160::from_low_u64_be(0x5001)
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct DappPrecompile<R>(PhantomData<R>);
+
+impl<R> PrecompileSet for DappPrecompile<R>
+where
+    R::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo + Decode,
+    <R::Call as Dispatchable>::Origin: From<Option<R::AccountId>>,
+    R: pallet_dapps_staking::Config + pallet_evm::Config,
+    R::Call: From<pallet_dapps_staking::Call<R>>,
+{
+    fn execute(
+        address: H160,
+        input: &[u8],
+        target_gas: Option<u64>,
+        context: &Context,
+    ) -> Option<Result<PrecompileOutput, ExitError>> {
+        match address {
+            a if a == precompile_address() => Some(DappsStakingWrapper::<R>::execute(
+                input, target_gas, context,
+            )),
+            _ => None,
+        }
+    }
+}
+
+pub type Precompiles = DappPrecompile<TestRuntime>;
+
+impl pallet_evm::Config for TestRuntime {
+    type FeeCalculator = ();
+    type GasWeightMapping = ();
+    type CallOrigin = EnsureAddressRoot<AccountId>;
+    type WithdrawOrigin = EnsureAddressNever<AccountId>;
+    type AddressMapping = AccountId;
+    type Currency = Balances;
+    type Event = Event;
+    type Runner = pallet_evm::runner::stack::Runner<Self>;
+    type Precompiles = DappPrecompile<TestRuntime>;
+    type ChainId = ();
+    type OnChargeTransaction = ();
+    type BlockGasLimit = ();
+    type BlockHashMapping = pallet_evm::SubstrateBlockHashMapping<Self>;
+    type FindAuthor = ();
+}
+
+parameter_types! {
+    pub const MinimumPeriod: u64 = 5;
+}
+impl pallet_timestamp::Config for TestRuntime {
+    type Moment = u64;
+    type OnTimestampSet = ();
+    type MinimumPeriod = MinimumPeriod;
+    type WeightInfo = ();
+}
+
+#[derive(PartialEq, Eq, Copy, Clone, Encode, Decode, Debug, scale_info::TypeInfo)]
+pub enum MockSmartContract<AccountId> {
+    Evm(sp_core::H160),
+    Wasm(AccountId),
+}
+
+impl<AccountId> Default for MockSmartContract<AccountId> {
+    fn default() -> Self {
+        MockSmartContract::Evm(H160::repeat_byte(0x00))
+    }
+}
+
+// impl MockSmartContract<AccountId> {
```

Comment:
You have some commented code in the file which should probably be removed.

General comment is that I wish mocks were more configurable in Substrate.
So much code duplication :(

Perhaps we should overengineer the problem and write a macro that generates mock :rofl:

Reply:
I will now go through the whole code again and clean the comments.
I also do not like the code duplication in the mock.
One approach might be to have this precompile as submodule for pallet-dapps-staking and then we could merge the tests and use the same mock